From 5a6f304f39c24c1a2c3023fbfda81b0042354c3f Mon Sep 17 00:00:00 2001
From: Nitin Rawat <quic_nitirawa@quicinc.com>
Date: Mon, 14 Jul 2025 13:23:36 +0530
Subject: [PATCH] scsi: ufs: ufs-qcom: Enable QUnipro Internal Clock Gating

[ Backport of 5a6f304f39c2 from 6.16 ]

Enable Unipro internal clock gating via DL_VS_CLK_CFG,
PA_VS_CLK_CFG_REG and DME_HW_CGC_EN. Power-saving feature: driver
continues if enabling fails.

[ Adapted for OnePlus 6.1.141 OKI: no ufshcd_dme_rmw(), use
  ufshcd_dme_get()/ufshcd_dme_set(). BIT(9) is free (cycle mask 0xFF). ]

Signed-off-by: Nitin Rawat <quic_nitirawa@quicinc.com>
---
diff --git a/drivers/ufs/host/ufs-qcom.c b/drivers/ufs/host/ufs-qcom.c
--- a/drivers/ufs/host/ufs-qcom.c
+++ b/drivers/ufs/host/ufs-qcom.c
@@ -347,12 +347,42 @@ static int ufs_qcom_power_up_sequence(struct ufs_hba *hba)
  */
 static void ufs_qcom_enable_hw_clk_gating(struct ufs_hba *hba)
 {
+	u32 reg;
+	int err;
+
+	/* Enable UTP internal clock gating */
 	ufshcd_writel(hba,
 		ufshcd_readl(hba, REG_UFS_CFG2) | REG_UFS_CFG2_CGC_EN_ALL,
 		REG_UFS_CFG2);
 
 	/* Ensure that HW clock gating is enabled before next operations */
 	ufshcd_readl(hba, REG_UFS_CFG2);
+
+	/* Enable Unipro internal clock gating */
+	err = ufshcd_dme_get(hba, UIC_ARG_MIB(DL_VS_CLK_CFG), &reg);
+	if (err)
+		goto out;
+	reg |= DL_VS_CLK_CFG_MASK;
+	err = ufshcd_dme_set(hba, UIC_ARG_MIB(DL_VS_CLK_CFG), reg);
+	if (err)
+		goto out;
+
+	err = ufshcd_dme_get(hba, UIC_ARG_MIB(PA_VS_CLK_CFG_REG), &reg);
+	if (err)
+		goto out;
+	reg |= PA_VS_CLK_CFG_REG_MASK;
+	err = ufshcd_dme_set(hba, UIC_ARG_MIB(PA_VS_CLK_CFG_REG), reg);
+	if (err)
+		goto out;
+
+	err = ufshcd_dme_get(hba, UIC_ARG_MIB(DME_VS_CORE_CLK_CTRL), &reg);
+	if (err)
+		goto out;
+	reg |= DME_VS_CORE_CLK_CTRL_DME_HW_CGC_EN;
+	err = ufshcd_dme_set(hba, UIC_ARG_MIB(DME_VS_CORE_CLK_CTRL), reg);
+out:
+	if (err)
+		dev_err(hba->dev, "hw clk gating enabled failed\n");
 }
 
 static int ufs_qcom_hce_enable_notify(struct ufs_hba *hba,
diff --git a/drivers/ufs/host/ufs-qcom.h b/drivers/ufs/host/ufs-qcom.h
--- a/drivers/ufs/host/ufs-qcom.h
+++ b/drivers/ufs/host/ufs-qcom.h
@@ -126,6 +126,15 @@ enum {
 /* bit and mask definitions for DME_VS_CORE_CLK_CTRL attribute */
 #define DME_VS_CORE_CLK_CTRL_CORE_CLK_DIV_EN_BIT		BIT(8)
 #define DME_VS_CORE_CLK_CTRL_MAX_CORE_CLK_1US_CYCLES_MASK	0xFF
+#define DME_VS_CORE_CLK_CTRL_DME_HW_CGC_EN			BIT(9)
+
+/* bit and mask definitions for PA_VS_CLK_CFG_REG attribute */
+#define PA_VS_CLK_CFG_REG	0x9004
+#define PA_VS_CLK_CFG_REG_MASK	GENMASK(8, 0)
+
+/* bit and mask definitions for DL_VS_CLK_CFG attribute */
+#define DL_VS_CLK_CFG		0xA00B
+#define DL_VS_CLK_CFG_MASK	GENMASK(9, 0)
 
 static inline void
 ufs_qcom_get_controller_revision(struct ufs_hba *hba,
